From acf5b86bb4d97b4a14731e7c196996a8e30a5397 Mon Sep 17 00:00:00 2001
From: Avery Pennarun <apenwarr@gmail.com>
Date: Wed, 3 Sep 2014 02:01:05 -0400
Subject: [PATCH] Return nonzero exit code if terminate_on_error is set.

Under some conditions, hostapd would abort on error but return a successful
exit code.

Example steps:
- hostapd starts and runs, but core dumps.
- this leaves mon.wlan0 and wlan0 interfaces in a funny state.
- upon next startup, can't initialize wlan0:
   nl80211: Failed to set channel (freq=2412): -16 (Device or resource busy)
   Could not set channel for kernel driver
   Interface initialization failed
- this leaves terminate_on_error nonzero because interface didn't init.
- however, the code then ignores terminate_on_error and just returns 0
  anyway.
- a babysitter looking for a successful exit would then not restart it.
---
 hostapd/main.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/hostapd/main.c b/hostapd/main.c
index 4f17688..53fec46 100644
--- a/hostapd/main.c
+++ b/hostapd/main.c
@@ -729,7 +729,7 @@ int main(int argc, char *argv[])
 		goto out;
 	}
 
-	ret = 0;
+	ret = interfaces.terminate_on_error;
 
  out:
 	hostapd_global_ctrl_iface_deinit(&interfaces);
-- 
2.1.0.rc2.206.gedb03e5

